[examples] Add how to package a system package using Conan (#3978)
conan-io/docs: examples/tools/system/package_manager.rst

.. _examples_tools_system_package_manager:

> System Packages

Reviewer: I think I'd use a more descriptive title, like ``Wrapping system requirements in a Conan package``.

Then, it would be great to discuss a bit more the problem/solution:

- Describe the problem that system_requirements by default execute
- At the end, describe how this approach avoids system_requirements() to be skipped and not run, if not needing to build

It would also be good, in a small subsection, to discuss the possibility of running the ``Apt()`` commands inside ``build()`` and cite pros and cons compared with the Conan wrapper recipe.

Author: > It would also be good, in a small subsection, to discuss the possibility of running the Apt() commands inside build() and cite pros and cons compared with the Conan wrapper recipe.

I would prefer keeping it in a next and separate PR, because there is a specific issue: https://github.com/conan-io/docs/issues/3975 So we can have a smaller PR here.

Author: @memsharded I updated the title on the commit [79f9fe1](https://github.com/conan-io/docs/pull/3978/commits/79f9fe196a15c7af54081998ebb60573f28f379d) and added a note about system requirements running always on the commit [4e4246e](https://github.com/conan-io/docs/pull/3978/commits/4e4246e0a93e386c3e47f9904fb7a47df8acae50)

===============================================

Conan can manage system packages, allowing you to install platform-specific dependencies easily.
This is useful when you need to install platform-specific system packages.
For example, you may need to install a package that provides a specific driver or graphics library that only works on a specific platform.

Conan provides a way to install system packages using the :ref:`system package manager<conan_tools_system_package_manager>` tool.

In this example, we are going to explore the steps needed to package a system library and what is needed to consume it in a Conan package.
For this illustration, we are going to package a simple C++ application that uses the `ncurses <https://invisible-island.net/ncurses/>`_ library.

Please, first clone the sources to recreate this project. You can find them in the
`examples2 repository <https://github.com/conan-io/examples2>`_ on GitHub:

.. code-block:: bash

    $ git clone https://github.com/conan-io/examples2.git
    $ cd examples2/examples/tools/system/package_manager/

> You may find the following tree structure:

Reviewer 2: ```suggestion
You will find the following tree structure:
```

.. code-block:: text

    .
    ├── conanfile.py
    └── consumer
        ├── CMakeLists.txt
        ├── conanfile.py
        └── ncurses_version.cpp

The **conanfile.py** file is the recipe that packages the ncurses library.
Finally, the **consumer** directory contains a simple C++ application that uses the ncurses library, we will visit it later.

> When packaging a pre-built system library, we do not need to build the project from source, but only to install the

Reviewer 2: ```suggestion
When packaging a pre-built system library, we do not need to build the project from source, only install the
```

system library and package its information.
In this case, we are going to check the **conanfile.py** file that packages the ncurses library first:

.. code-block:: python

    from conan import ConanFile
    from conan.tools.system import package_manager
    from conan.tools.gnu import PkgConfig
    from conan.errors import ConanInvalidConfiguration

    required_conan_version = ">=2.0"

    class SysNcursesConan(ConanFile):
        name = "ncurses"
        version = "system"
        description = "A textual user interfaces that work across a wide variety of terminals"
        topics = ("curses", "terminal", "toolkit")
        homepage = "https://invisible-mirror.net/archives/ncurses/"
        license = "MIT"
        package_type = "shared-library"
        settings = "os", "arch", "compiler", "build_type"

        def package_id(self):
            self.info.clear()

        def validate(self):
            if self.settings.os != "Linux":

>                 raise ConanInvalidConfiguration(f"{self.ref} wraps a system package only supported by Linux.")

Reviewer 3: looks like ncurses is also available for macOS with brew: https://formulae.brew.sh/formula/ncurses

Author: Updated on the commit [0be72f1](https://github.com/conan-io/docs/pull/3978/commits/0be72f1be4fa9dbf7fb982b9a89d35f4c3bcc7f0). Also added support for FreeBSD:

https://github.com/conan-io/examples2/pull/177#issuecomment-2630822064

            yum.install(["ncurses-devel"], update=True, check=True)

            apt = package_manager.Apt(self)
            apt.install(["libncurses-dev"], update=True, check=True)

            pacman = package_manager.PacMan(self)
            pacman.install(["ncurses"], update=True, check=True)

            zypper = package_manager.Zypper(self)
            zypper.install(["ncurses"], update=True, check=True)

        def package_info(self):
            self.cpp_info.bindirs = []
            self.cpp_info.includedirs = []
            self.cpp_info.libdirs = []

            pkg_config = PkgConfig(self, 'ncurses')
            pkg_config.fill_cpp_info(self.cpp_info, is_system=True)

In this **conanfile.py** file, we are using the :ref:`system package manager<conan_tools_system_package_manager>` tool
to install the ncurses library based on different package managers, under the
:ref:`system_requirements<reference_conanfile_methods_system_requirements>` method. It's important to note that the
``system_requirements`` method is called always, when building, or even if the package is already installed.
This is useful to ensure that the package is installed in the system.

> Each package may vary the package name used to install the ncurses library, so we need to check the package manager

Reviewer 2: ```suggestion
Each package manager may vary the package name used to install the ncurses library, so we need to check the package manager
```

documentation to find the correct package name first.

> Another important details is the **package_info** method. In this method, we are using the

Reviewer 2: ```suggestion
Another important detail is the **package_info** method. In this method, we are using the
```

        generators = "CMakeDeps", "CMakeToolchain"
        package_type = "application"
        exports_sources = "CMakeLists.txt", "ncurses_version.cpp"

        def requirements(self):
            if self.settings.os == "Linux":
                self.requires("ncurses/system")

        def layout(self):
            cmake_layout(self)

        def build(self):
            cmake = CMake(self)
            cmake.configure()
            cmake.build()

            self.run(os.path.join(self.build_folder, "ncurses_version"), env="conanrun")

The recipe is simple. It requires the **ncurses** package we just created and uses the **CMake** tool to build the application.
Once the application is built, it runs the **ncurses_version** application, so we can check the executable output as its result.

The **ncurses_version.cpp** file is a simple C++ application that uses the ncurses library to print the ncurses version:

.. code-block:: cpp

    #include <cstdlib>
    #include <cstdio>
    #include <ncurses.h>

    int main(int argc, char *argv[]) {
        printf("Conan 2.x Examples - Installed NCurses version: %s\n", curses_version());
        return EXIT_SUCCESS;

>     }

Reviewer 3: I think that I would try to "use" ncurses for the example, not just printing the version, it could be something as simple as moving the cursor to the center of the screen and printing a message.

Author: Moving the cursor did not work for me, because I have it disabled on my terminal. Instead, I updated the example to show a colored message (white background, blue text).

https://github.com/conan-io/examples2/pull/177#issuecomment-2631136269

.. _examples_tools_system_package_manager:

Wrapping system requirements in a Conan package
===============================================

> Conan not only manages C and C++ dependencies, but it can also manage system packages.

Reviewer 3: I think it's better to start with a positive statement
```suggestion
Conan can manage system packages, allowing you to install platform-specific dependencies easily.
```

> This is useful when you need to install strictly platform-specific system packages.

Reviewer 3: ```suggestion
This is useful when you need to install platform-specific system packages.
```

For example, you may need to install a package that provides a specific driver or graphics library that only works on a specific platform.

Conan provides a way to install system packages using the :ref:`system package manager<conan_tools_system_package_manager>` tool.

In this example, we are going to explore the steps needed to package a system library and what is needed to consume it in a Conan package.
For this illustration, we are going to package a simple C++ application that uses the `ncurses <https://invisible-island.net/ncurses/>`_ library.

> 

Author: Updated on the commit d8f0b89c

Please, first clone the sources to recreate this project. You can find them in the
`examples2 repository <https://github.com/conan-io/examples2>`_ in GitHub:

.. code-block:: bash

    $ git clone https://github.com/conan-io/examples2.git
    $ cd examples2/examples/tools/system/package_manager/

You may find the following tree structure:

.. code-block:: text

    .
    ├── ci_test_example.py
    ├── conanfile.py
    └── consumer
        ├── CMakeLists.txt
        ├── conanfile.py
        └── ncurses_version.cpp

The **ci_test_example.py** file is a script consumed by the CI to check if the example is running without errors,

> you should not worry about this file.

Reviewer 3: We don't usually mention this file in the tutorials, I think it's better to omit it.

Author: Done, omitted from the documentation page.

The **conanfile.py** file is the recipe that packages the ncurses library.
Finally, the **consumer** directory contains a simple C++ application that uses the ncurses library, we will visit it later.

> When packaging a pre-built system library, we do not need to build a project from the source, but only to install the

Reviewer 3: ```suggestion
When packaging a pre-built system library, we do not need to build the project from source, but only to install the
```

.. _examples_tools_system_package_manager:

Wrapping system requirements in a Conan package
===============================================

Conan can manage system packages, allowing you to install platform-specific dependencies easily.
This is useful when you need to install platform-specific system packages.
For example, you may need to install a package that provides a specific driver or graphics library that only works on a specific platform.

Conan provides a way to install system packages using the :ref:`system package manager<conan_tools_system_package_manager>` tool.

> In this example, we are going to explore the steps needed to package a system library and what is needed to consume it in a Conan package.

Reviewer: To create a wrapper package around a system library. Note that the package will not contain the binary artifacts, it will just manage to check/install them calling ``system_requirements()`` and the respective Apt/Yum/... system package managers.

Author: Done, please see the commit 5215cde6

> For this illustration, we are going to create a Conan package to wrap the system `ncurses <https://invisible-island.net/ncurses/>`_

Reviewer: ``For this illustration`` => In this example

Author: Done, please, see the commit 7bcd1abd

requirement and then show how to use this requirement in an application.

Please, first clone the sources to recreate this project. You can find them in the
`examples2 repository <https://github.com/conan-io/examples2>`_ on GitHub:

.. code-block:: bash

    $ git clone https://github.com/conan-io/examples2.git
    $ cd examples2/examples/tools/system/package_manager/

You will find the following tree structure:

.. code-block:: text

    .
    ├── conanfile.py
    └── consumer
        ├── CMakeLists.txt
        ├── conanfile.py
        └── ncurses_version.c

> The **conanfile.py** file is the recipe that packages the ncurses library.

Reviewer: ```suggestion
The ``conanfile.py`` file is the recipe that wraps the ncurses system library.
```

Finally, the **consumer** directory contains a simple C application that uses the ncurses library, we will visit it later.

> When packaging a pre-built system library, we do not need to build the project from source, only install the

Reviewer: ```suggestion
When wrapping a pre-built system library, we do not need to build the project from source, only install the
```

As a result of this command, you should be able to see the **ncurses** library installed in your system, in case not been installed yet.

Now, let's check the **consumer** directory. This directory contains a simple C application that uses the ncurses library.

The **conanfile.py** file in the **consumer** directory is:

.. code-block:: python

    from conan import ConanFile
    from conan.tools.build import can_run
    from conan.tools.cmake import cmake_layout, CMake
    import os

    class AppNCursesVersionConan(ConanFile):
        settings = "os", "compiler", "build_type", "arch"
        generators = "CMakeDeps", "CMakeToolchain"
        package_type = "application"
        exports_sources = "CMakeLists.txt", "ncurses_version.c"

        def requirements(self):
            if self.settings.os in ["Linux", "Macos", "FreeBSD"]:
                self.requires("ncurses/system")

        def layout(self):
            cmake_layout(self)

        def build(self):
            cmake = CMake(self)
            cmake.configure()
            cmake.build()

            if self.settings.os in ["Linux", "Macos", "FreeBSD"]:

>                 self.run(os.path.join(self.build_folder, "ncurses_version"), env="conanrun")

Reviewer 3: Do we want to execute the example in the build method? Wouldn't it be better to execute from the executable location via the command line?  I find this a bit unexpected

Reviewer: I suggested it as a way to make it simpler for both users and CI and writing the documentation. A single ``conan build .`` will work for CI and for all users, if not, we need to start dealing with:
- Different paths and layouts in Windows and Linux
- Potential activation of environment different in Windows and Linux (maybe not necessary here)

Reviewer 3: I see your point about keeping it simple, but I think running the executable manually has more educational value and is more coherent with the rest of the docs. As a middle ground, how about printing the path to the executable at the end of `build()`, like:

```
self.output.info(f"Application built. You can run it invoking: '{os.path.join(self.build_folder, 'ncurses_version')}'")
```

Reviewer: Ok, good for me then, can be done as @czoido suggests

Author: Done, please, see the commit: b758f39f

It also reflects the current state of the example: https://github.com/conan-io/examples2/pull/177/commits/240f393dd2fb2ac310961bdfa684b4c308909b0a

        int max_y, max_x;
        char message [256] = {0};

        initscr();

        start_color();
        init_pair(1, COLOR_BLUE, COLOR_WHITE);
        getmaxyx(stdscr, max_y, max_x);

        snprintf(message, sizeof(message), "Conan 2.x Examples - Installed ncurses version: %s\n", curses_version());
        attron(COLOR_PAIR(1));
        mvprintw(max_y / 2, max_x / 2 - (strlen(message) / 2), "%s", message);
        attroff(COLOR_PAIR(1));

        refresh();

        return EXIT_SUCCESS;
    }

The **CMakeLists.txt** file is a simple CMake file that builds the **ncurses_version** application:

.. code-block:: cmake

    cmake_minimum_required(VERSION 3.15)
    project(ncurses_version C)

    find_package(Curses CONFIG REQUIRED)

    add_executable(${PROJECT_NAME} ncurses_version.c)
    target_link_libraries(${PROJECT_NAME} PRIVATE Curses::Curses)

The CMake target **Curses::Curses** is provided by the **ncurses** package we just created.
The information about libraries and include directories is now available in the **cpp_info** object, as we filled it using the **PkgConfig** tool.

Now, let's build the application:

.. code-block:: bash

    $ cd consumer/

>     $ conan build . --name=ncurses-version --version=0.1.0 --build-folder=build

Reviewer 3: ```suggestion
    $ conan build . --name=ncurses-version --version=0.1.0
```

The  `--build-folder` argument does not exist in `conan build` for Conan 2

        attron(COLOR_PAIR(1));
        mvprintw(max_y / 2, max_x / 2 - (strlen(message) / 2), "%s", message);
        attroff(COLOR_PAIR(1));

        refresh();

        return EXIT_SUCCESS;
    }

The **CMakeLists.txt** file is a simple CMake file that builds the **ncurses_version** application:

.. code-block:: cmake

    cmake_minimum_required(VERSION 3.15)
    project(ncurses_version C)

    find_package(Curses CONFIG REQUIRED)

    add_executable(${PROJECT_NAME} ncurses_version.c)
    target_link_libraries(${PROJECT_NAME} PRIVATE Curses::Curses)

The CMake target **Curses::Curses** is provided by the **ncurses** package we just created. It follows the official CMake module for `FindCurses <https://cmake.org/cmake/help/latest/module/FindCurses.html>`_.
The information about libraries and include directories is now available in the **cpp_info** object, as we filled it using the **PkgConfig** tool.

Now, let's build the application:

.. code-block:: bash

    $ cd consumer/
    $ conan build . --name=ncurses-version --version=0.1.0

After building the application, it should be executed automatically, so you may see its output:

> 

Reviewer 3: This changed, they will see message to help them execute the application.
